### PR TITLE
System C# analyzer to detect fields or locals that could be declared with more specific types

### PR DESCRIPTION
**Pull request by JohnHamby**
This change set adds another System C# analyzer as a test of IOperation in analyzer actions.

No fixes or additions to the analyzer infrastructure were necessary, but there are a couple of changes to language infrastructure for generating IOperation nodes:
  1) The C# member semantic model always failed to produce a bound node for a member initializer. (This was a bug.)
  2) The VB member semantic model produces a wrapper node for a member initializer rather than the bound node for the initializer expression itself, so turning that node into an IOperation needs to unwrap the expression.

The VB and C# engines have asymmetrical treatment of member initializers--VB produces a bound node for the "= expr" that wraps the node for "expr", while C# does not. This seems like a spurious difference.

@mavasani @srivatsn @heejaechang  for the analyzer.
@gafter @AlekseyTs for the language infrastructure.

### CODE REVIEW

**Comment by JohnHamby** [comment]
Adding @VSadov @cston

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs [hunk 1]
```diff
@@ -1067,7 +1073,7 @@ private ImmutableArray<BoundNode> GuardedAddBoundTreeAndGetBoundNodeFromMap(CSha
 
             if (bound != null)
             {
-                alreadyInTree = _guardedNodeMap.ContainsKey(bound.Syntax);
+                alreadyInTree = _guardedNodeMap.ContainsKey(AsNodeMapKey(bound.Syntax));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: This change doesn't feel right. The bound node is associated with a particular syntax, that syntax should be used to cache the node. If the node is associated with wrong syntax, it would be better to change binding to use the right syntax for the node.

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs [hunk 2]
```diff
@@ -1054,7 +1054,13 @@ private ImmutableArray<BoundNode> GuardedGetBoundNodesFromMap(CSharpSyntaxNode n
         {
             Debug.Assert(_nodeMapLock.IsWriteLockHeld || _nodeMapLock.IsReadLockHeld);
             ImmutableArray<BoundNode> result;
-            return _guardedNodeMap.TryGetValue(node, out result) ? result : default(ImmutableArray<BoundNode>);
+            return _guardedNodeMap.TryGetValue(AsNodeMapKey(node), out result) ? result : default(ImmutableArray<BoundNode>);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: I think this is not the right place to change the lookup key. This method should return bound nodes really associated with the given syntax node. I think the adjustments should be made in a different place. Either we should change binding to create a special bound node associated with EqualsValue syntax, or we should change initializer semantic model to never request a node for EqualsValue syntax.

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs (hunk 1 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: I agree with @AlekseyTs. In general, it is going to be difficult to ensure consistency between bound tree structures in VB and C#, it could be in place for performance reasons. Additionally, we could also break lowering passes in doing so. The bound node tree structure can also undergo any changes in future - its not a public API, but an implementation detail. The right approach would be to handle these differences at the layer that maps syntax/bound nodes to IOperations. We should write more unit tests to uncover these differences, and prevent any public facing regressions.

Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: @mavasani it is not possible to handle this entirely at the mapping of bound nodes to IOperations, because the MemberSemanticModel currently always returns null if asked to bind any field initializer code block. This is a bug that needs fixing. @AlekseyTs has suggested (offline) introducing a bound node to represent the field initialization explicitly, and I'll try that.

File: src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs [hunk 3]
```diff
@@ -181,20 +181,22 @@ private BoundNode BindEqualsValue(Binder binder, EqualsValueClauseSyntax equalsV
...
+                        return new BoundFieldInitializer(equalsValue, fieldSymbol, binder.BindVariableOrAutoPropInitializer(equalsValue, fieldType, diagnostics));
                     }
 
                 case SymbolKind.Property:
                     {
-                        var propertyType = ((PropertySymbol)this.MemberSymbol).Type;
-                        return binder.BindVariableOrAutoPropInitializer(equalsValue, propertyType, diagnostics);
+                        var propertySymbol = (PropertySymbol)this.MemberSymbol;
+                        var propertyType = propertySymbol.Type;
+                        return new BoundPropertyInitializer(equalsValue, propertySymbol, binder.BindVariableOrAutoPropInitializer(equalsValue, propertyType, diagnostics));
```
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs]: I added BoundPropertyInitializer for symmetry. It isn't possible to exercise this path with the patterns of the analyzer I was writing, but this code path is apparently live--when I inserted an exception throw here a couple of tests failed. I'm a bit uncomfortable knowing that this code can be reached by some means other than analyzing code blocks for member initializers--and not knowing precisely what those means are--but all tests pass with this change.

**Comment by JohnHamby** [comment]
The change set now includes creating a bound node for an "= expr" in a field initializer. This seems to work OK, but I'm not 100% sure I understand the implications.

File: src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs [hunk 4]
```diff
@@ -181,20 +181,22 @@ private BoundNode BindEqualsValue(Binder binder, EqualsValueClauseSyntax equalsV
             {
                 case SymbolKind.Field:
                     {
-                        var enumField = this.MemberSymbol as SourceEnumConstantSymbol;
+                        var fieldSymbol = (FieldSymbol)this.MemberSymbol;
+                        var enumField = fieldSymbol as SourceEnumConstantSymbol;
                         if ((object)enumField != null)
                         {
                             return binder.BindEnumConstantInitializer(enumField, equalsValue.Value, diagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs]: What node does this return?

File: src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs]: BoundPropertyInitializer doesn't exist because the initializer is actually initializes the backing field and compiler creates BoundFieldInitializer node instead.

File: src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs [hunk 5]
```diff
@@ -181,20 +181,22 @@ private BoundNode BindEqualsValue(Binder binder, EqualsValueClauseSyntax equalsV
...
                         }
 
-                        var fieldType = ((FieldSymbol)this.MemberSymbol).GetFieldType(binder.FieldsBeingBound);
-                        return binder.BindVariableOrAutoPropInitializer(equalsValue, fieldType, diagnostics);
+                        var fieldType = (fieldSymbol).GetFieldType(binder.FieldsBeingBound);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs]: Extra parentheses.

File: src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs]: Rather than using BoundFieldInitializer/BoundPropertyInitializer in here, I would introduce a new bound node, BoundEqualsValue derived from BoundNode with only one property - Value (we don't really need the target symbol), and would change signature of this function to return BoundEqualsValue node. This way all initializers will be handled uniformly here. For example, why are we not changing ```case SymbolKind.Parameter```? Is just because we don't have a unit-test for IOperation that goes through it? Perhaps we should add one? It feels like we should have unit-tests for IOperation covering every single branch in this function.

File: src/Compilers/VisualBasic/Portable/Binding/MemberSemanticModel.vb [hunk 6]
```diff
@@ -784,14 +784,25 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
                 Case GetOperationOptions.Parent
-                    Return TryCast(summary.LowestBoundNodeOfSyntacticParent, IOperation)
+                    result = summary.LowestBoundNodeOfSyntacticParent
                 Case Else
-                    Return TryCast(summary.LowestBoundNode, IOperation)
+                    result = summary.LowestBoundNode
             End Select
+
+            ' Screen out bound nodes that aren't appropriate as IOperations.
+            If result IsNot Nothing Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/MemberSemanticModel.vb]: Are we handling EqualsValue nodes in other contexts? In parameter's declaration, for example?

File: src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs (hunk 3 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs]: OK, will do. I was trying to follow the shape of the what the VB model does, while also leveraging the BoundFieldInitializer node that was already in existence, and was going to deal with parameter default values if the treatment of fields was acceptable. I agree that a single uniform treatment is better.

File: src/Compilers/VisualBasic/Portable/Binding/MemberSemanticModel.vb (hunk 6 above)
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/Binding/MemberSemanticModel.vb]: Parameter default values and explicit Enum member values were not covered--thanks. The treatment of these is slightly less systematic in VB than in C# because not all VB initializers produce expressions.

File: src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs (hunk 4 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs]: Something that needed to be wrapped... thanks.

File: src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs (hunk 3 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs]: As @AlekseyTs correctly observed, test coverage was not adequate. It's more thorough now.

File: src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb [hunk 7]
```diff
@@ -97,12 +97,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                         Debug.Assert(initializer IsNot Nothing)
                         If initializer.Kind = SyntaxKind.EqualsValue Then
                             Dim enumSymbol = DirectCast(Me.MemberSymbol, SourceEnumConstantSymbol)
-                            boundInitializer = binder.BindFieldAndEnumConstantInitializer(enumSymbol, DirectCast(initializer, EqualsValueSyntax), isEnum:=True, diagnostics:=diagnostics, constValue:=Nothing)
+                            boundInitializer = WrapWithEqualsValue(binder.BindFieldAndEnumConstantInitializer(enumSymbol, DirectCast(initializer, EqualsValueSyntax), isEnum:=True, diagnostics:=diagnostics, constValue:=Nothing), initializer)
```
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb]: This wrapping occurs in a couple of specific places rather than in all cases because some cases produce initializers that are not expressions.

File: src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb [hunk 8]
```diff
@@ -136,6 +136,14 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             Return boundInitializer
         End Function
 
+        Private Function WrapWithEqualsValue(boundValue As BoundExpression, value As VisualBasicSyntaxNode) As BoundNode
+            If boundValue IsNot Nothing AndAlso boundValue.Kind <> BoundKind.BadExpression Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb]: Bad expression is still an expression, I think we should remove ```AndAlso boundValue.Kind <> BoundKind.BadExpression``` condition.

**Comment by AlekseyTs** [comment]
LGTM, module the BadExpression handling.

File: src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb (hunk 7 above)
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb]: My thinking above is incorrect--the cases that produce things other than expressions also need to be covered. I'll add some tests for things like "As New" declarations, and expect to have to rework this code again.

File: src/Compilers/VisualBasic/Test/Semantic/Diagnostics/OperationAnalyzerTests.vb [hunk 9]
```diff
@@ -415,5 +415,254 @@ End Class
...
+End Enum
+
+Class C1
+    Public Sub New (a As Integer, b As Integer, c As Integer)
+    End Sub
+
+    Public F1 As C1 = New C1(c:=16, a:=17, b:=18)
+    Public F2 As New C1(16, 17, 18)
+    Public F3(16) As Integer
+    Public F4(17) As Integer                          ' The upper bound specification is not presently treated as a code block. This is suspect.
```
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Test/Semantic/Diagnostics/OperationAnalyzerTests.vb]: This ought to be detected and presently is not. The logic to find code blocks in declarations is not finding array upper bound specifications. This issue is outside the scope of this change set.

File: src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb [hunk 10]
```diff
@@ -133,7 +133,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                     Throw ExceptionUtilities.UnexpectedValue(Me.MemberSymbol.Kind)
             End Select
 
-            Return boundInitializer
+            Dim expressionInitializer = TryCast(boundInitializer, BoundExpression)
+            Return If(expressionInitializer IsNot Nothing AndAlso expressionInitializer.Kind <> BoundKind.BadExpression, New BoundEqualsValue(initializer, expressionInitializer), boundInitializer)
```
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb]: Yes, testing for bad expressions here seems unsound. Without this, there is a failure in a flow analysis test that includes a lambda definition in the default value expression of a parameter. Teaching the flow analysis logic about BoundEqualsValue nodes is probably necessary if checking for bad expressions is to be removed.

Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb]: Producing proxy nodes for all initializers (including those that aren't expressions) caused numerous test failures--some in flow analysis tests and some in various editor tests. Wrapping expression cases is sufficient for extracting IOperations, and don't lead to test failures in other areas.

**Comment by JohnHamby** [comment]
Is anyone interested in reviewing the could-have-more-specific-type analyzer itself?

File: src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb (hunk 10 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb]: @JohnHamby It feels suspicious that BoundEqualsValue node flows to flow analysis only for bad expressions. Perhaps the only flow analysis test we have for default parameter value is testing bad expression. It should be easy to clone and modify that test to test non-bad expression. I can help you with that.

File: src/Compilers/Core/CodeAnalysisTest/Diagnostics/CouldHaveMoreSpecificTypeAnalyzer.cs [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+                                // Track explicit assignments.
+                                operationBlockContext.RegisterOperationAction(
+                                   (operationContext) =>
+                                   {
+                                       IAssignmentExpression assignment = (IAssignmentExpression)operationContext.Operation;
+                                       AssignTo(assignment.Target, localsSourceTypes, fieldsSourceTypes, assignment.Value);
+                                   },
+                                   OperationKind.AssignmentExpression,
+                                   OperationKind.CompoundAssignmentExpression,
+                                   OperationKind.IncrementExpression);
```
Comment (srivatsn) [inline src/Compilers/Core/CodeAnalysisTest/Diagnostics/CouldHaveMoreSpecificTypeAnalyzer.cs]: If there was code like Foo(ref field1) would the assignment to field1 be covered by the AssignmentExpression?

**Comment by srivatsn** [comment]
The analyzer looks fine to me.

File: src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb (hunk 10 above)
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb]: I think it more likely that flow analysis aborts if given a bad expression, but that feeding flow analysis a BoundEqualsValue that isn't bad but that has a bad Value causes the analysis to occur (and get an assertion failure) for a bad lambda.

There may be a few places that need to be taught to skip through BoundEqualsValue nodes. The first approach--adding some extra entries in the map from syntax nodes to bound nodes--was actually more robust in that it could not introduce problems for other kinds of analysis.

File: src/Compilers/Core/CodeAnalysisTest/Diagnostics/CouldHaveMoreSpecificTypeAnalyzer.cs (hunk 11 above)
Comment (JohnHamby) [inline src/Compilers/Core/CodeAnalysisTest/Diagnostics/CouldHaveMoreSpecificTypeAnalyzer.cs]: All ref arguments are treated as assigned to, by the action registered for InvocationExpression immediately below.

File: src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb (hunk 10 above)
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb]: The issue turned out to be that flow analysis was looking for a BoundExpression and instead found a BoundEqualsValue. Going through the BoundEqualsValue to a BoundExpression solved the problem.

However, I have no idea how many other lurking issues of this sort might remain. The more I think about it, the more I'm worried that introducing a placeholder node that any client must recognize and skip through is not the best behavior. Mapping multiple syntax nodes to one bound node in a node map may feel a little unclean, but it completely avoids the possibility of introducing problems like this.

Also, VB has already been doing something much like this. I found this code in MemberSemanticModel.vb:

If Not _guardedNodeMap.ContainsKey(bindingRoot) Then
                ' Generally 'bindingRoot' is supposed to be found in node map at this point,
                ' but it will not happen in some scenarios such as for field or property 
                ' initializers, let's add it to prevent re-binding 

                Debug.Assert(bindingRoot.Kind = SyntaxKind.FieldDeclaration OrElse
                             bindingRoot.Kind = SyntaxKind.PropertyStatement OrElse
                             bindingRoot.Kind = SyntaxKind.Parameter OrElse
                             bindingRoot.Kind = SyntaxKind.EnumMemberDeclaration OrElse
                             bindingRoot Is Me.Root AndAlso Me.IsSpeculativeSemanticModel)

                _guardedNodeMap.Add(bindingRoot, ImmutableArray.Create(Of BoundNode)(boundRoot))

File: src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs [hunk 12]
```diff
@@ -288,6 +288,15 @@ protected BoundNode VisitAlways(BoundNode node)
         private BoundNode VisitWithStackGuard(BoundNode node)
         {
             var expression = node as BoundExpression;
+            if (expression == null)
+            {
+                var equalsValue = node as BoundEqualsValue;
+                if (equalsValue != null)
+                {
+                    expression = equalsValue.Value;
```
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs]: I'm not 100% sure this statement is reachable for C#, but the corresponding logic in VB was necessary.

**Comment by JohnHamby** [comment]
An slightly different approach would be to filter out the BoundEqualsValue nodes before they can escape the SemanticModels. For example, both the VB and C# MemberSemanticModels go through a GetBoundNodes method, and BoundEqualsValue nodes could be filtered out there (or in Get(Upper/Lower)BoundNode methods). This would preserve the invariant of mapping only one syntax node to a given bound node, while not requiring clients to filter out placeholder nodes.

This would cause me less worry, but the decision belongs to @AlekseyTs  or @gafter.

File: src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs [hunk 13]
```diff
@@ -161,6 +161,11 @@ protected enum RegionPlace { Before, Inside, After };
             bool trackExceptions = false)
         {
             Debug.Assert(node != null);
+            var equalsValue = node as BoundEqualsValue;
+            if (equalsValue != null)
+            {
+                node = equalsValue.Value;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs]: This is not the right way to handle BoundEqualsValue node,the class should simply override VisitEqualsValue method.

**Comment by AlekseyTs** [comment]
This is not the first time we add a new bound node and it won't be the last one. We shouldn't be afraid to do this. Usually, unit-tests are very good at pointing to the code that is supposed to be adjusted once a new node is added. Obviously, it is better to be proactive and go through all tree walkers/rewriters in the code base and make appropriate changes to them.
About the cache map, I would prefer to keep it consistent and avoid situations when keys point to the set of nodes that aren't point back to the keys. If we were talking about fixing a bug close to a release, temporarily patching the map would be acceptable as a risk mitigation measure, but I am not sure we are this point here.

File: src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs [hunk 14]
```diff
@@ -2563,6 +2563,12 @@ public override BoundNode VisitImplicitReceiver(BoundImplicitReceiver node)
             return null;
         }
 
+        public override BoundNode VisitEqualsValue(BoundEqualsValue node)
```
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs]: I copied the form of some other overriding Visit*** methods. Hopefully this is close to correct.

**Comment by AlekseyTs** [comment]
LGTM
